src/lib.rs

```diff
@@ -0,0 +1,309 @@
+            // TODO: Limit access
+            ensure_root(origin)?;
+            // Ensure chain is whitelisted
+            if let Some(mut counter) = Chains::get(&dest_id) {
+                // Increment counter and store
+                counter.recv += 1;
+                Chains::insert(&dest_id, counter.clone());
+                Self::deposit_event(RawEvent::AssetTransfer(dest_id, counter.recv, to, token_id, metadata));
+                Ok(())
+            } else {
+                Err(Error::<T>::ChainNotWhitelisted)?
+            }
+        }
+
+        // TODO: Should use correct amount type
+        pub fn transfer(origin, to: T::AccountId, amount: u32) -> DispatchResult {
+            ensure_root(origin)?;
+            let source: T::AccountId = <EndowedAccount<T>>::get();
+            T::Currency::transfer(&source, &to, amount.into(), AllowDeath)?;
+            Ok(())
+        }
+    }
+}
+
+/// Main module declaration.
+/// Here we should include non-state changing public funcs
+impl<T: Trait> Module<T> {
+    pub fn is_validator(who: &T::AccountId) -> bool {
+        Self::validators(who)
+    }
+
+    fn initialize_validators(validators: &[T::AccountId]) {
+        if !validators.is_empty() {
+            for v in validators {
+                <Validators<T>>::insert(v, true);
+            }
+        }
+    }
+
+    /// Note: Existence of proposal must be checked before calling
+    fn vote_for(
+        who: T::AccountId,
+        mut votes: ProposalVotes<T::AccountId, T::Hash>,
+    ) -> DispatchResult {
+        //let mut prop = <Proposals<T>>::get((deposit_id, call.clone())).unwrap();
+        if !votes.votes_for.contains(&who) {
+            votes.votes_for.push(who.clone());
+            <Votes<T>>::insert(votes.hash, votes.clone());
+            Self::deposit_event(RawEvent::VoteFor(votes.hash, who.clone()));
+
+            if votes.votes_for.len() == <ValidatorThreshold>::get() as usize {
+                Self::finalize_transfer(votes)?
+            } else if votes.votes_for.len() > <ValidatorThreshold>::get() as usize {
+                Err(Error::<T>::ProposalAlreadyComplete)?
+            }
+            Ok(())
+        } else {
+            Err(Error::<T>::ValidatorAlreadyVoted)?
+        }
+    }
+
+    /// Note: Existence of proposal must be checked before calling
+    fn vote_against(
+        who: T::AccountId,
+        mut votes: ProposalVotes<T::AccountId, T::Hash>,
+    ) -> DispatchResult {
+        // let mut prop = <Proposals<T>>::get((deposit_id, call.clone())).unwrap();
+        if !votes.votes_against.contains(&who) {
+            votes.votes_against.push(who.clone());
+            <Votes<T>>::insert(votes.hash, votes.clone());
+            Self::deposit_event(RawEvent::VoteAgainst(votes.hash, who.clone()));
+
+            if votes.votes_against.len() > <ValidatorThreshold>::get() as usize {
```

Comment:
I believe the logic is wrong here – this should rather be: votes against > validator set size - threshold